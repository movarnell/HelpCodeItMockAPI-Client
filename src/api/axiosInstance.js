// src/api/axiosInstance.js
import axios from 'axios';

const axiosInstance = axios.create({
  baseURL: 'http://localhost:5100', // Adjust if necessary
});

axiosInstance.interceptors.request.use(
    (config) => {
      const token = localStorage.getItem('token');
      console.log('Token from localStorage:', token);
      if (token) {
        config.headers['Authorization'] = `Bearer ${token}`;
      } else {
        console.warn('No token found in localStorage');
      }
      return config;
    },
    (error) => {
      return Promise.reject(error);
    }
  );

export default axiosInstance;