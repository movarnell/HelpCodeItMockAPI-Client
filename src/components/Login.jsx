// src/components/Login.jsx
import React, { useState } from 'react';
import axiosInstance from '../api/axiosInstance';
import { useNavigate, Link } from 'react-router-dom';
import logo from '../assets/Mock It.png';

function Login() {
  const [username, setUsername] = useState('');
  const [password, setPassword] = useState('');
  const [error, setError] = useState('');
  const navigate = useNavigate();

  const handleSubmit = async (e) => {
    e.preventDefault();
    try {
      const response = await axiosInstance.post('/auth/login', {
        username,
        password,
      });
      localStorage.setItem('token', response.data.token);
      navigate('/dashboard');
    } catch (err) {
      setError(err.response.data.message || 'Login failed');
    }
  };

  return (
    <div className="flex items-center justify-center h-screen bg-gray-100 rounded">
      <div className="w-full max-w-md p-8 space-y-3 bg-white rounded shadow">
        {/* <h1 className="text-4xl font-bold text-center">Mock API</h1> */}
        <img src={logo} alt="logo" className="w-1/2 mx-auto" />
        <h1 className="text-2xl font-bold text-center">Login</h1>
        {error && <p className="text-red-500">{error}</p>}
        <form onSubmit={handleSubmit} className="space-y-6">
          <div>
            <label className="block text-sm">Username</label>
            <input
              type="text"
              className="w-full px-4 py-2 border rounded-md"
              value={username}
              onChange={(e) => setUsername(e.target.value)}
              required
            />
          </div>
          <div>
            <label className="block text-sm">Password</label>
            <input
              type="password"
              className="w-full px-4 py-2 border rounded-md"
              value={password}
              onChange={(e) => setPassword(e.target.value)}
              required
            />
          </div>

          <button className="w-full px-4 py-2 font-bold text-white bg-blue-500 rounded-md hover:bg-blue-600">
            Login
          </button>
          <p className="text-sm text-center">
            Don't have an account?{' '}
            <Link to="/register" className="text-blue-500 hover:underline">
              Register
            </Link>
            <br />
            This app is currently free to use.
            <br />If you have any issues please let us know using our <a href="https://www.helpcodeit.com/contact" className="text-blue-500 hover:underline">contact form!</a>
          </p>
        </form>
      </div>
    </div>
  );
}

export default Login;