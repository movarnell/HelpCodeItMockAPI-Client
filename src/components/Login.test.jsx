import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { BrowserRouter, useNavigate } from 'react-router-dom';
import Login from './Login';
import axiosInstance from '../api/axiosInstance';
import { act } from 'react-dom/test-utils';

jest.mock('../api/axiosInstance');
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: jest.fn(),
}));

describe('Login Component', () => {
  const mockNavigate = jest.fn();

  beforeEach(() => {
    useNavigate.mockReturnValue(mockNavigate);
    mockNavigate.mockClear();
  });

  const renderComponent = async () => {
    await act(async () => {
      render(
        <BrowserRouter>
          <Login />
        </BrowserRouter>
      );
    });
  };

  test('renders login form', async () => {
    await renderComponent();
    expect(screen.getByLabelText(/username/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/password/i)).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /login/i })).toBeInTheDocument();
  });

  test('handles form submission', async () => {
    axiosInstance.post.mockResolvedValueOnce({ data: { token: 'fake-token' } });
    await renderComponent();

    await act(async () => {
      fireEvent.change(screen.getByLabelText(/username/i), { target: { value: 'testuser' } });
      fireEvent.change(screen.getByLabelText(/password/i), { target: { value: 'password123' } });
      fireEvent.click(screen.getByRole('button', { name: /login/i }));
    });

    await waitFor(() => {
      expect(axiosInstance.post).toHaveBeenCalledWith('/auth/login', {
        username: 'testuser',
        password: 'password123'
      });
      expect(mockNavigate).toHaveBeenCalledWith('/dashboard');
    });
  });

  test('displays error message on login failure', async () => {
    axiosInstance.post.mockRejectedValueOnce({ response: { data: { message: 'Invalid credentials' } } });
    await renderComponent();

    await act(async () => {
      fireEvent.change(screen.getByLabelText(/username/i), { target: { value: 'testuser' } });
      fireEvent.change(screen.getByLabelText(/password/i), { target: { value: 'wrongpassword' } });
      fireEvent.click(screen.getByRole('button', { name: /login/i }));
    });

    await waitFor(() => {
      expect(screen.getByText('Invalid credentials')).toBeInTheDocument();
    });
  });
});