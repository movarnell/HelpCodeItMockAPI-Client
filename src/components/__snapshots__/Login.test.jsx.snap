// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Component matches snapshot 1`] = `
<div
  className="flex items-center justify-center h-screen bg-gray-100 rounded"
>
  <div
    className="w-full max-w-md p-8 space-y-3 bg-white rounded shadow"
  >
    <img
      alt="logo"
      className="w-1/2 mx-auto"
      src="test-file-stub"
    />
    <h1
      className="text-2xl font-bold text-center"
    >
      Login
    </h1>
    <form
      className="space-y-6"
      onSubmit={[Function]}
    >
      <div>
        <label
          className="block text-sm"
          htmlFor="username"
        >
          Username
        </label>
        <input
          className="w-full px-4 py-2 border rounded-md"
          id="username"
          onChange={[Function]}
          required={true}
          type="text"
          value=""
        />
      </div>
      <div>
        <label
          className="block text-sm"
          htmlFor="password"
        >
          Password
        </label>
        <input
          className="w-full px-4 py-2 border rounded-md"
          id="password"
          onChange={[Function]}
          required={true}
          type="password"
          value=""
        />
      </div>
      <button
        className="w-full px-4 py-2 font-bold text-white bg-blue-500 rounded-md hover:bg-blue-600"
      >
        Login
      </button>
      <p
        className="text-sm text-center"
      >
        Don't have an account?
         
        <a
          className="text-blue-500 hover:underline"
          href="/register"
          onClick={[Function]}
        >
          Register
        </a>
        <br />
        This app is currently free to use.
        <br />
        If you have any issues please let us know using our 
        <a
          className="text-blue-500 hover:underline"
          href="https://www.helpcodeit.com/contact"
        >
          contact form!
        </a>
      </p>
    </form>
  </div>
</div>
`;
